worlds/yugioh06/__init__.py

```diff
@@ -0,0 +1,538 @@
+                                   challenge, {challenge: lid, challenge + " Complete": None})
+            entrance = Entrance(self.player, challenge, challenge_region)
+            entrance.access_rule = (lambda unlock: lambda state: state.has(unlock + " Unlock", self.player))(challenge)
+            challenge_region.exits.append(entrance)
+            entrance.connect(region)
+            self.multiworld.regions.append(region)
+
+    def generate_early(self):
+        if self.options.structure_deck.current_key == "none":
+            boosters = draft_boosters
+            if self.options.campaign_opponents_shuffle.value:
+                opponents = tier_1_opponents
+            else:
+                opponents = draft_opponents
+        else:
+            boosters = booster_packs
+            opponents = tier_1_opponents
+
+        if self.options.structure_deck.current_key == "random_deck":
+            self.options.structure_deck.value = self.random.choice([0, 1, 2, 3, 4, 5])
+        for item in self.options.start_inventory:
+            if item in opponents:
+                self.starting_opponent = item
+            if item in boosters:
+                self.starting_booster = item
+        if not self.starting_opponent:
+            self.starting_opponent = self.multiworld.random.choice(opponents)
+        self.multiworld.push_precollected(self.create_item(self.starting_opponent))
+        if not self.starting_booster:
+            self.starting_booster = self.multiworld.random.choice(boosters)
+        self.multiworld.push_precollected(self.create_item(self.starting_booster))
+        banlist = self.options.banlist.value
+        self.multiworld.push_precollected(self.create_item(Banlist_Items.get(banlist)))
+        challenge = list((Limited_Duels | Theme_Duels).keys())
+        noc = len(challenge) - max(self.options.third_tier_5_campaign_boss_challenges.value
+                                   if self.options.third_tier_5_campaign_boss_unlock_condition.value == 1 else 0,
+                                   self.options.fourth_tier_5_campaign_boss_challenges.value
+                                   if self.options.fourth_tier_5_campaign_boss_unlock_condition.value == 1 else 0,
+                                   self.options.final_campaign_boss_challenges.value
+                                   if self.options.final_campaign_boss_unlock_condition.value == 1 else 0,
+                                   self.options.number_of_challenges.value,
+                                   91 if hasattr(self.multiworld, "generation_is_fake") else 0)
+
+        self.random.shuffle(challenge)
+        excluded = self.options.exclude_locations.value.intersection(challenge)
+        prio = self.options.priority_locations.value.intersection(challenge)
+        normal = [e for e in challenge if e not in excluded and e not in prio]
+        total = list(excluded) + normal + list(prio)
+        self.removed_challenges = total[:noc]
```

Comment:
i assume you're not shuffling to remove the excluded challenges first?
if so why not shuffle each individual list before combining? or at least shuffle normal

Reply:
challenge is shuffled beforehand a few lines above